In [2]:
import math
import numpy as np
import decimal

# Prostory jmen a obor platnosti proměnných:
- Upravte funkci nalézající prvočísla tak, aby ukládala již spočítané prvočísla do listu v defaultním argumentu funkce. 


In [15]:
def Jeprvocislo(a):
    for i in range(2,a):
        if a%i == 0:
            return False
        return True

Jeprvocislo(6)

#cisla=[]
def prvocisla(n,cisla=[]):
    count = 0
    x = 1
    prime=0
    while count < n:
        prime=Jeprvocislo(x)
        if prime == True:
            cisla.append(x)
            count += 1
        x += 1
    return cisla

prvocisla(10)

[3, 5, 7, 9, 11, 13, 15, 17, 19, 21]


# Jupyter magics:
- použijte funkci %%writefile a pomocí ní (a kopie funkcí z posledního DU) vytvořte .py soubor sloužící jako modul mající tři funkce viz minulý DU
- Použijte magic příkazy %time %timeit %%time a %%timeit k měření času běhu různých částí kódu (například funkcí z minulého DU).


In [16]:
%%writefile soubor.py

import numpy as np
import math
a=100
n=10
def odmocnina (a,n):
    x = [0] * n
    x[0] = a
    for i in range(n-1):  
        #print(i)
        x[i+1] = (a/x[i] + x[i])/2
        #print(x) 
    #print(x[n-1])
    return x[n-1]
odmocnina(a,n)
np.sqrt(100)

def Uhelniky(n):
    an,bn = 1,1
    
    if n == 0:
        return S
    else:
        for i in range(0,n+1):
            #print(i)
            if i == 0:
                an = 1
                bn = 1
                vn = math.sqrt(an**2 - ((bn/2)**2))
                #print("vn", vn)
                S = 6*(2**i)*bn*(vn/2)
                #print("S=",S)
            if i > 0:
                #print("vn", vn)
                b2n = math.sqrt((bn/2)**2 + (1-vn)**2)
                #print("b2n = ", b2n)
                v2n = math.sqrt(an**2 - (b2n/2)**2)
                #print("v2n=",v2n)
                S = 6*(2**i)*b2n*v2n/2 
                #print("S=",S)
                i = i + 1
                bn = b2n
                vn = v2n
    return S


def pi3(n):
    #a = [0]* (n+1)
    #a[1] = (1/2) * (1/2**3)
    a1 = (1/2) * (1/2**3)
    s = 1/3* a1
    for i in range(2,n+1):
        a2 = a1 * ((2*i-3)/(2*i))*(1/(2**2))
        a1 = a2
        s = s +  1/(2*i+1)*a2
    pi = 12 * (-np.sqrt(3)/8 + 1/2 - s)
    return pi

Overwriting soubor.py


In [17]:
def Uhelniky(n):
    an,bn = 1,1
    
    if n == 0:
        return S
    else:
        for i in range(0,n+1):
            #print(i)
            if i == 0:
                an = 1
                bn = 1
                vn = math.sqrt(an**2 - ((bn/2)**2))
                #print("vn", vn)
                S = 6*(2**i)*bn*(vn/2)
                #print("S=",S)
            if i > 0:
                #print("vn", vn)
                b2n = math.sqrt((bn/2)**2 + (1-vn)**2)
                #print("b2n = ", b2n)
                v2n = math.sqrt(an**2 - (b2n/2)**2)
                #print("v2n=",v2n)
                S = 6*(2**i)*b2n*v2n/2 
                #print("S=",S)
                i = i + 1
                bn = b2n
                vn = v2n
    return S

In [18]:
def pi3(n):
    #a = [0]* (n+1)
    #a[1] = (1/2) * (1/2**3)
    a1 = (1/2) * (1/2**3)
    s = 1/3* a1
    for i in range(2,n+1):
        a2 = a1 * ((2*i-3)/(2*i))*(1/(2**2))
        a1 = a2
        s = s +  1/(2*i+1)*a2
    pi = 12 * (-np.sqrt(3)/8 + 1/2 - s)
    return pi

In [19]:
def odmocnina (a,n):
    x = [0] * n
    x[0] = a
    for i in range(n-1):  
        #print(i)
        x[i+1] = (a/x[i] + x[i])/2
        #print(x) 
    #print(x[n-1])
    return x[n-1]


In [21]:
%time odmocnina(567432,10)
%time Uhelniky(111)
%time pi3(111111)

Wall time: 0 ns
Wall time: 0 ns
Wall time: 40 ms


3.1415926535897936


# Práce s řetězci:

- Použíjte funkce z minulého úkolu a v cyklu spočtěte $\pi$ s různým $n$. Pomocí f-strings vypište výsledky ve formátu "n = 5, pi = 3.14159, počet správných cifer = 5".


In [22]:
from decimal import getcontext, Decimal as D

In [31]:
def Uhelniky(n):
    an,bn = 1,1
    
    if n == 0:
        return S
    else:
        for i in range(0,n+1):
            #print(i)
            if i == 0:
                an = 1
                bn = 1
                vn = math.sqrt(an**2 - ((bn/2)**2))
                #print("vn", vn)
                S = 6*(2**i)*bn*(vn/2)
                #print("S=",S)
            if i > 0:
                #print("vn", vn)
                b2n = math.sqrt((bn/2)**2 + (1-vn)**2)
                #print("b2n = ", b2n)
                v2n = math.sqrt(an**2 - (b2n/2)**2)
                #print("v2n=",v2n)
                S = 6*(2**i)*b2n*v2n/2 
                #print("S=",S)
                i = i + 1
                bn = b2n
                vn = v2n
    return S    


In [57]:
math_pi=math.pi
for i in range(5,10):
    placesU = math.floor(abs(math.log10(math_pi-Uhelniky(i))))
    print(f"Uhelniky(n): n={i}, pi={Uhelniky(i)}, pocet spravnych cifer {placesU}")
    

Uhelniky(n): n=5, pi=3.1410319508905102, pocet spravnych cifer 3
Uhelniky(n): n=6, pi=3.1414524722854624, pocet spravnych cifer 3
Uhelniky(n): n=7, pi=3.1415576079118575, pocet spravnych cifer 4
Uhelniky(n): n=8, pi=3.1415838921483186, pocet spravnych cifer 5
Uhelniky(n): n=9, pi=3.1415904632280505, pocet spravnych cifer 5



# Generátory a iterátory:

- Napište generátor (funkci), který vytváří Fibonacciho posloupnost a vrátí prvních $n$ členů.


In [65]:
def Fibonaci(n):
    count=0
    i=0
    j=1
    members=[]
    #members.append(i)
    #members.append(1)
    if n==1:
        members.append(i)
        return members
    else:
        while count < n:
            next=i+j
            print(i)
            members.append(i)
            i=j
            j=next
            count+=1
    return members


In [66]:
Fibonaci(7)

0
1
1
2
3
5
8


[0, 1, 1, 2, 3, 5, 8]


# List comprehension:
- Vytvořte seznam čísel od 1 do 10 pomocí list comprehension.
- Vytvořte seznam druhých mocnin čísel od 1 do 10 pomocí list comprehension.
- Vytvořte seznam prvočísel od 1 do 100 pomocí list comprehension. Hint (použijte funkci all() na generátor testující dělitelnost).


In [1]:
list = [i for i in range(1,11)]
list
mocniny = [i**2 for i in range(1,11)]
mocniny
primes = [i for i in range(2,101) if all( i % y != 0 for y in range(2,i))]
primes

[2,
 3,
 5,
 7,
 11,
 13,
 17,
 19,
 23,
 29,
 31,
 37,
 41,
 43,
 47,
 53,
 59,
 61,
 67,
 71,
 73,
 79,
 83,
 89,
 97]


# Práce se soubory:
- Pomocí %%writefile vytvořte soubor s nějakým textem.
- Otevřete soubor pro čtení a vypište jeho obsah.
- Otevřete jiný soubor pro zápis a napište do něj textový řetězec.
- Pomocí f-strings zapište do souboru tabulku čísel od 1 do 10, jejich druhých mocnin a jejich druhých odmocnin.

In [100]:
%%writefile text.txt
Lorem ipsum dolor sit amet, consectetur adipiscing elit. 
Vivamus eget gravida erat, quis ultrices massa.
Suspendisse commodo ante augue, et cursus nibh consectetur vel.
Nulla mattis convallis nibh a scelerisque. Nullam at porttitor arcu.
Maecenas eleifend ex sapien tristique, a sagittis nulla fringilla.

Writing text.txt


In [102]:
soubor = open("text.txt", "r")
print(soubor.read())
soubor.close()

Lorem ipsum dolor sit amet, consectetur adipiscing elit. 
Vivamus eget gravida erat, quis ultrices massa.
Suspendisse commodo ante augue, et cursus nibh consectetur vel.
Nulla mattis convallis nibh a scelerisque. Nullam at porttitor arcu.
Maecenas eleifend ex sapien tristique, a sagittis nulla fringilla.



In [109]:
with open("zapis.csv", "w") as soubor:
    soubor.write("Nejaky text \n")
    for i in range(1, 11):
        soubor.write(f"{i*i}, {np.sqrt(i)} \n")


In [111]:
%less zapis.csv

Nejaky text 
1, 1.0 
4, 1.4142135623730951 
9, 1.7320508075688772 
16, 2.0 
25, 2.23606797749979 
36, 2.449489742783178 
49, 2.6457513110645907 
64, 2.8284271247461903 
81, 3.0 
100, 3.1622776601683795 


# Domácí úkol:
Pomocí vlastních tříd a dědičnosti vytvořte "soubor tříd" = nástroj pro derivaci funkcí.

Konkrétněji:
- Vytvořte třídu **Operator**
    - Třída bude mít metody:
        - `__init__()` - inicializuje operátor (zde stačí `pass`)
        - `__str__()` - vrací řetězec reprezentující operátor
        - `__repr__()` - vrací řetězec reprezentující operátor
- Vytvořte třídu **BinaryOperator**, která bude dědit od třídy Operator
    - Třída bude mít atributy:
        - `levy_argument` - levý operand
        - `pravy_argument` - pravý operand
    - Třída bude mít metody:
        - `__init__()` - inicializuje operátor, přiřadí argumenty
        - `__str__()` - vrací řetězec reprezentující operátor
        - `__repr__()` - vrací řetězec reprezentující operátor
        - `diff()` - vrací derivaci operátoru
- Vytvořte třídu **UnaryOperator**, která bude dědit od třídy Operator
    - Třída bude mít atributy:
        - `argument` - argument funkce 
    - Třída bude mít metody:
        - `__init__()` - inicializuje operátor, přiřadí argument
        - `__str__()` - vrací řetězec reprezentující operátor
        - `__repr__()` - vrací řetězec reprezentující operátor
        - `diff()` - vrací derivaci operátoru

Pro počítání derivací použijte standardní pravidla:
- derivace součtu je součet derivací
- pravidlo pro derivaci součinu
- derivace funkce je derivovaná funkce násobená derivovaným argumentem

Pro **BinaryOperator** a **UnaryOperator** implementujte také metody implementující standardní operátory:
- `__add__()` - implementuje operátor `+`
- `__sub__()` - implementuje operátor `-`
- `__mul__()` - implementuje operátor `*`

Pro **BinaryOperator** implementujte dvě dceřiné třídy:
- **Addition** - implementuje operátor `+`
- **Multiplication** - implementuje operátor `*`

Pro **UnaryOperator** implementujte tyto dceřiné třídy:
- **exp** - implementuje funkci `exp`
- **sin** - implementuje funkci `sin`
- **cos** - implementuje funkci `cos`
- **ln** - implementuje funkci `ln`
- **frc** - implementuje funkci `1/x`
- **identity** - implementuje identitu `x`
- **cst** - implementuje konstantu `c`
- **mocnina** - implementuje mocninu `x^n`

U všech funckí implementujte adekvátně metodu `diff()` a `__str__()`, tak aby se funkce vypisovaly v přehledné podobě.

Ozkoušejte na následujících funkcích:
- $sin(x + cos(x*x)) + 1$
- $x^2 + 2x + 1 + 2x^2$
- $x^2 + 2x + exp(x)$

In [76]:
# kostra k doplnění, místa označená TODO doplňte

class Operator:
    def __init__(self):
        pass
    
    def __str__(self):
        return type(self).__name__
    
    def __repr__(self):
        return self.__str__()
    
    def diff(self):
        pass
    
    def __add__(self, other):
        # TODO (použijte Addition)
        return Addition(self,other)

    
    def __sub__(self, other):
        # TODO (použijte Addition a Multiplication se zápornou konstantu)
        return Addition(self, Multiplication(-1,other))
    
    def __mul__(self, other):
        # TODO (použijte Multiplication)
        return Multiplication(self,other)
    
class BinaryOperator(Operator):
    def __init__(self, levy_argument, pravy_argument):
        self.levy_argument = levy_argument
        self.pravy_argument = pravy_argument
        
class UnaryOperator(Operator):
    def __init__(self, argument):
        self.argument = argument
    
    def __str__(self):
        return type(self).__name__ + "(" + str(self.argument) + ")"
        
class Addition(BinaryOperator):
    def __str__(self):
        # TODO vypíšeme něco jako (levy_argument)+(pravy_argument)
        return "(" + str(self.levy_argument) + ")+(" + str(self.pravy_argument) + ")"

    def diff(self):
        # TODO
        return Addition(self.levy_argument.diff(), self.pravy_argument.diff())
    
    def simplify(self):
        if (self.levy_argument == 0 and isinstance(self.pravy_argument,cst)):
            return self.pravy_argument
        if (self.pravy_argument == 0 and isinstance(self.levy_argument,cst)):
            return self.levy_argument
        if (self.levy_argument == 0 and self.pravy_argument == 0):
            return cst(0)
        if (isinstance(self.levy_argument,cst) and isinstance(self.pravy_argument,cst)):
            return (self.levy_argument.value + self.pravy_argument.value)
    

class Multiplication(BinaryOperator):
    def __str__(self):
        # TODO vypíšeme něco jako (levy_argument)*(pravy_argument)
        return "(" + str(self.levy_argument) + ")*(" + str(self.pravy_argument) + ")"
    def diff(self):
        # TODO
        return Addition(Multiplication(self.levy_argument.diff(),self.pravy_argument), Multiplication(self.levy_argument,self.pravy_argument.diff()))
    
    def simplify(self):
        if (self.levy_argument == 0 or self.pravy_argument == 0):
            return cst(0)
        if (self.levy_argument == self.pravy_argument):
            return mocnina(self.levy_argument,2)
        if (self.levy_argument == 1):
            return self.pravy_argument
        if (self.pravy_argument == 1):
            return self.levy_argument
        if(isinstance(self.levy_argument, cst) and isinstance(self.pravy_argument, cst)):
            return self.levy_argument.value * self.pravy_argument.value

class cst(UnaryOperator):
    def __init__(self, value):
        self.value = value
    def __str__(self):
        # TODO vypíšeme něco jako value
        return str(self.value)
    def diff(self):
        # TODO
        return cst(0)
    
class identity(UnaryOperator):
    def __init__(self):
        pass
    def __str__(self):
        # TODO vypíšeme něco jako x  
        var = "x" 
        return str(var)
    def diff(self):
        # TODO
        return cst(1)
    
class mocnina(UnaryOperator):
    def __init__(self, argument, exponent):
        self.argument = argument
        self.exponent = exponent
    def __str__(self):
        # TODO vypíšeme něco jako argument^exponent
        return str(self.argument) + "^" + str(self.exponent)
    def diff(self):
        # TODO
        return Multiplication(self.exponent, mocnina(self.argument, self.exponent-1))
    
    def simplify(self):
        if (self.exponent == 0):
            return cst(1)
        if (self.exponent ==1):
            return self.argument

class sin(UnaryOperator):
    def __str__(self):
        return "sin(" + str(self.argument) + ")"

    def diff(self):
        # TODO
        return Multiplication(cos(self.argument),self.argument.diff())

class cos(UnaryOperator):
    def __str__(self):
        return "cos(" + str(self.argument) + ")"
    def diff(self):
        # TODO
        return Multiplication(-1,Multiplication(sin(self.argument),self.argument.diff()))

class exp(UnaryOperator):
    def __str__(self):
        return "(e^" + str(self.argument) + ")"
    def diff(self):
        # TODO
        return exp(self.argument)
    
class ln(UnaryOperator):
    def __str__(self):
        return "ln(" + str(self.argument) + ")"
    def diff(self):
        # TODO
        return Multiplication(frc(self.argument), self.argument.diff())

class frc(UnaryOperator):
    def __str__(self):
        # TODO vypíšeme něco jako 1/(argument)
        return "1/(" + str(self.argument) + ")"
    
    def diff(self):
        # TODO
        return Multiplication(-1, frc(exp(self.argument, 2)))


In [77]:
x = identity()
f = mocnina(x,2) + cst(2)*x + exp(x)
f.diff()

(((2)*(x^1))+(((0)*(x))+((2)*(1))))+((e^x))

In [78]:
x =identity()
f = x * 0
print(f.simplify())

0


In [79]:
# takto by se to mělo používat
x = identity()
# sin(x + cos(x*x)) + 1
f1 = sin(x + cos(x*x)) + cst(1)
print(f1)
print(f1.diff())

(sin((x)+(cos((x)*(x)))))+(1)
((cos((x)+(cos((x)*(x)))))*((1)+((-1)*((sin((x)*(x)))*(((1)*(x))+((x)*(1)))))))+(0)


můj výstup byl:

`sin(x + cos((x)*(x))) + 1`

`(cos(x + cos((x)*(x))))*(1 + ((-1)*(sin((x)*(x))))*((1)*(x) + (x)*(1))) + 0`

In [80]:
x = identity()
# x ^ 2 + 2x + 1 + 2x ^ 2
f2 = mocnina(x,2) + cst(2)*x + cst(1) + cst(2)*mocnina(x,2)
print(f2)
print(f2.diff())


(((x^2)+((2)*(x)))+(1))+((2)*(x^2))
((((2)*(x^1))+(((0)*(x))+((2)*(1))))+(0))+(((0)*(x^2))+((2)*((2)*(x^1))))


můj výstup byl:

`(x)^2 + (2)*(x) + 1 + (2)*((x)^2)`

`((2)*((x)^1))*(1) + (0)*(x) + (2)*(1) + 0 + (0)*((x)^2) + (2)*(((2)*((x)^1))*(1))`

In [81]:
x = identity()
# x ^ 2 + 2x + exp(x)
f3 = mocnina(x,2) + cst(2)*x + exp(x)
print(f3)
print(f3.diff())


((x^2)+((2)*(x)))+((e^x))
(((2)*(x^1))+(((0)*(x))+((2)*(1))))+((e^x))


můj výstup byl:

`(x)^2 + (2)*(x) + exp(x)`

`((2)*((x)^1))*(1) + (0)*(x) + (2)*(1) + (exp(x))*(1)`


# Bonusový úkol:
Doplňte ke všem třídám metodu `simplify()`, která provede zjednodušení výrazu. Například:
- součet s nulou vrátí druhý operand
- součin s nulou vrátí nulu
- mocnina s exponentem 0 vrátí 1
- mocnina s exponentem 1 vrátí první argument
- součín dvou stejných argumentů vrátí mocninu s exponentem 2
- součin konstant vrátí konstantu se součinem
- ...